executor/analyze.go

```diff
@@ -381,3 +398,401 @@ func (e *AnalyzeColumnsExec) buildStats() (hists []*statistics.Histogram, cms []
+			if err != nil {
+				return nil, nil, errors.Trace(err)
+			}
+		}
+	}
+	for i, collector := range collectors {
+		var hg *statistics.Histogram
+		if i == 0 && hasPKInfo > 0 {
+			hg, err = statistics.BuildColumn(e.ctx, int64(e.maxNumBuckets), e.pkInfo.ID, collector, &e.pkInfo.FieldType)
+		} else {
+			hg, err = statistics.BuildColumn(e.ctx, int64(e.maxNumBuckets), e.colsInfo[i-hasPKInfo].ID, collector, &e.colsInfo[i-hasPKInfo].FieldType)
+		}
+		if err != nil {
+			return nil, nil, errors.Trace(err)
+		}
+		hists = append(hists, hg)
+		cms = append(cms, collectors[i].CMSketch)
+	}
+	// collect index samples.
+	if hasIdxInfo > 0 {
+		// TODO: build index statistic result
+	}
+	return
+}
+
+func (e *AnalyzeFastExec) handleSampTasks(bo *tikv.Backoffer, needRebuild *bool, err *error, hists *[]*statistics.Histogram, cms *[]*statistics.CMSketch) {
+	defer func() {
+		if *err != nil || *needRebuild == true {
+			for _, ok := <-e.tasks; ok; _, ok = <-e.tasks {
+				// Do nothing.
+			}
+		}
+	}()
+	client := e.ctx.GetStore().(tikv.Storage).GetTiKVClient()
+	for {
+		task, ok := <-e.tasks
+		if !ok {
+			return
+		}
+
+		if task.isScan {
+			// TODO: handle scan task.
+		}
+
+		var tableID, minRowID, maxRowID int64
+		startKey, endKey := task.Location.StartKey, task.Location.EndKey
+		tableID, minRowID, *err = tablecodec.DecodeRecordKey(startKey)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		_, maxRowID, *err = tablecodec.DecodeRecordKey(endKey)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+
+		keys := make([][]byte, 0, task.SampSize)
+		for i := 0; i < int(task.SampSize); i++ {
+			randKey := rand.Int63n(maxRowID-minRowID) + minRowID
+			keys = append(keys, tablecodec.EncodeRowKeyWithHandle(tableID, randKey))
+		}
+
+		var resp *tikvrpc.Response
+		var rpcCtx *tikv.RPCContext
+		rpcCtx, *err = e.cache.GetRPCContext(bo, task.Location.Region)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		ctx := context.Background()
+		req := &tikvrpc.Request{
+			Type:        tikvrpc.CmdBatchGet,
+			RawBatchGet: &kvrpcpb.RawBatchGetRequest{Keys: keys},
+		}
+		resp, *err = client.SendRequest(ctx, rpcCtx.Addr, req, tikv.ReadTimeoutMedium)
+		if *err != nil {
+			*err = errors.Trace(*err)
+			return
+		}
+		regionErr, err1 := resp.GetRegionError()
+		if err1 != nil {
+			*err = err1
+			return
+		}
+		if regionErr != nil {
+			if err1 := bo.Backoff(tikv.BoRegionMiss, errors.New(regionErr.String())); err1 != nil {
+				*err = errors.Trace(err1)
+				return
+			}
+			*needRebuild = true
+			return
+		}
+
+		tmpHists, tmpCms, err1 := e.handleBatchGetResponse(resp.BatchGet)
+		if err1 != nil {
+			*err = errors.Trace(err1)
+			return
+		}
+		if len(*hists) == 0 {
+			*hists = tmpHists
+			*cms = tmpCms
+		} else {
+			for i := 0; i < len(*hists); i++ {
+				(*hists)[i], *err = statistics.MergeHistograms(e.ctx.GetSessionVars().StmtCtx, (*hists)[i], tmpHists[i], int(e.maxNumBuckets))
```

Comment:
`MergeHistograms` assumes that there is no overlap between the two histograms and this may not be true here. We can just build the histogram after all samples have got.